File: src/modules/modelManager.ts
/**
 * 🗃️ Model Manager Module
 * 
 * Downloads, caches, and manages local AI models with platform optimizations.
 * Handles model variants, quantizations, and automatic selection based on system capabilities.
 */

import { createHash } from 'crypto';
import { createReadStream, createWriteStream, existsSync, mkdirSync, statSync, unlinkSync, readFileSync, writeFileSync } from 'fs';
import { join, dirname } from 'path';
import { homedir } from 'os';
import { platformDetector, SystemCapabilities, ModelCompatibility } from './platformDetector';

export interface ModelInfo {
  name: string;
  displayName: string;
  description: string;
  size: number;           // Size in MB
  quantization: 'f16' | 'f32' | 'q4_0' | 'q4_1' | 'q5_0' | 'q5_1' | 'q8_0';
  architecture: 'llama' | 'mistral' | 'phi' | 'tinyllama' | 'other';
  contextLength: number;  // Maximum context length in tokens
  requiredMemory: number; // Minimum RAM needed in MB
  downloadUrl: string;
  checksum: string;       // SHA256 checksum for verification
  tags: string[];         // ['chat', 'instruct', 'code', etc.]
  license: string;
  createdBy: string;
}

export interface DownloadProgress {
  modelName: string;
  downloadedBytes: number;
  totalBytes: number;
  percentage: number;
  speed: number;          // bytes per second
  eta: number;            // estimated time remaining in seconds
  status: 'downloading' | 'verifying' | 'complete' | 'error';
  error?: string;
}

export interface ModelCache {
  path: string;
  size: number;
  checksum: string;
  downloadDate: Date;
  lastUsed: Date;
  useCount: number;
}

export class ModelManager {
  private modelsDir: string;
  private cacheFile: string;
  private downloadCallbacks: Map<string, (progress: DownloadProgress) => void> = new Map();
  
  constructor(customModelsDir?: string) {
    this.modelsDir = customModelsDir || join(homedir(), '.ethervoxai', 'models');
    this.cacheFile = join(this.modelsDir, 'cache.json');
    this.ensureDirectories();
  }

  /**
   * Get the default model catalog with popular GGML models
   */
  getDefaultModelCatalog(): ModelInfo[] {
    return [
      {
        name: 'tinyllama-1.1b-chat-q4',
        displayName: 'TinyLlama 1.1B Chat (Q4)',
        description: 'Lightweight model perfect for basic conversations and resource-constrained devices',
        size: 669,
        quantization: 'q4_0',
        architecture: 'tinyllama',
        contextLength: 2048,
        requiredMemory: 1024,
        downloadUrl: 'https://huggingface.co/TheBloke/TinyLlama-1.1B-Chat-v1.0-GGML/resolve/main/tinyllama-1.1b-chat-v1.0.q4_0.bin',
        checksum: 'placeholder_checksum_1',
        tags: ['chat', 'lightweight', 'embedded'],
        license: 'Apache-2.0',
        createdBy: 'TinyLlama Team'
      },
      {
        name: 'phi-2-2.7b-q4',
        displayName: 'Microsoft Phi-2 2.7B (Q4)',
        description: 'High-quality small model from Microsoft, great for ARM devices',
        size: 1600,
        quantization: 'q4_0',
        architecture: 'phi',
        contextLength: 2048,
        requiredMemory: 2048,
        downloadUrl: 'https://huggingface.co/microsoft/phi-2-ggml/resolve/main/phi-2.q4_0.bin',
        checksum: 'placeholder_checksum_2',
        tags: ['chat', 'instruct', 'arm-optimized'],
        license: 'MIT',
        createdBy: 'Microsoft Research'
      },
      {
        name: 'mistral-7b-instruct-v0.1-q4',
        displayName: 'Mistral 7B Instruct v0.1 (Q4)',
        description: 'Excellent instruction-following model with strong performance',
        size: 4100,
        quantization: 'q4_0',
        architecture: 'mistral',
        contextLength: 4096,
        requiredMemory: 5120,
        downloadUrl: 'https://huggingface.co/TheBloke/Mistral-7B-Instruct-v0.1-GGML/resolve/main/mistral-7b-instruct-v0.1.q4_0.bin',
        checksum: 'placeholder_checksum_3',
        tags: ['chat', 'instruct', 'general'],
        license: 'Apache-2.0',
        createdBy: 'Mistral AI'
      },
      {
        name: 'llama2-7b-chat-q4',
        displayName: 'Llama 2 7B Chat (Q4)',
        description: 'Popular general-purpose chat model from Meta',
        size: 3900,
        quantization: 'q4_0',
        architecture: 'llama',
        contextLength: 4096,
        requiredMemory: 4896,
        downloadUrl: 'https://huggingface.co/TheBloke/Llama-2-7B-Chat-GGML/resolve/main/llama-2-7b-chat.q4_0.bin',
        checksum: 'placeholder_checksum_4',
        tags: ['chat', 'general', 'popular'],
        license: 'Custom (Llama 2)',
        createdBy: 'Meta'
      },
      {
        name: 'llama2-13b-chat-q4',
        displayName: 'Llama 2 13B Chat (Q4)',
        description: 'High-capability model for systems with sufficient memory',
        size: 7300,
        quantization: 'q4_0',
        architecture: 'llama',
        contextLength: 4096,
        requiredMemory: 8192,
        downloadUrl: 'https://huggingface.co/TheBloke/Llama-2-13B-Chat-GGML/resolve/main/llama-2-13b-chat.q4_0.bin',
        checksum: 'placeholder_checksum_5',
        tags: ['chat', 'general', 'high-performance'],
        license: 'Custom (Llama 2)',
        createdBy: 'Meta'
      }
    ];
  }

  /**
   * Get recommended models based on system capabilities
   */
  async getRecommendedModels(): Promise<ModelInfo[]> {
    const capabilities = await platformDetector.getCapabilities();
    const catalog = this.getDefaultModelCatalog();
    
    const recommended: ModelInfo[] = [];
    
    for (const model of catalog) {
      const compatibility = await platformDetector.checkModelCompatibility(
        model.name,
        model.size,
        model.requiredMemory
      );
      
      if (compatibility.isCompatible) {
        recommended.push(model);
      }
    }
    
    // Sort by performance tier preference
    recommended.sort((a, b) => {
      const perfOrder = ['ultra', 'high', 'medium', 'low'];
      const aScore = capabilities.performanceTier === 'ultra' ? a.size : -a.size;
      const bScore = capabilities.performanceTier === 'ultra' ? b.size : -b.size;
      return bScore - aScore;
    });
    
    return recommended;
  }

  /**
   * Download a model with progress tracking
   */
  async downloadModel(
    modelName: string,
    onProgress?: (progress: DownloadProgress) => void
  ): Promise<string> {
    const catalog = this.getDefaultModelCatalog();
    const modelInfo = catalog.find(m => m.name === modelName);
    
    if (!modelInfo) {
      throw new Error(`Model '${modelName}' not found in catalog`);
    }

    // Check if already downloaded
    const modelPath = join(this.modelsDir, `${modelName}.bin`);
    if (existsSync(modelPath)) {
      const isValid = await this.verifyModel(modelPath, modelInfo.checksum);
      if (isValid) {
        console.log(`✅ Model '${modelName}' already downloaded and verified`);
        return modelPath;
      } else {
        console.log(`⚠️ Model '${modelName}' checksum mismatch, re-downloading...`);
        unlinkSync(modelPath);
      }
    }

    console.log(`📥 Downloading model: ${modelInfo.displayName} (${modelInfo.size}MB)`);
    
    // Register progress callback
    if (onProgress) {
      this.downloadCallbacks.set(modelName, onProgress);
    }

    try {
      // For now, we'll simulate the download since we don't have actual URLs
      // In production, this would use fetch() or https.get() for actual downloads
      await this.simulateDownload(modelInfo, modelPath, onProgress);
      
      // Update cache
      await this.updateCache(modelName, modelPath, modelInfo);
      
      console.log(`✅ Successfully downloaded: ${modelInfo.displayName}`);
      return modelPath;
      
    } catch (error) {
      if (onProgress) {
        onProgress({
          modelName,
          downloadedBytes: 0,
          totalBytes: modelInfo.size * 1024 * 1024,
          percentage: 0,
          speed: 0,
          eta: 0,
          status: 'error',
          error: error instanceof Error ? error.message : 'Unknown error'
        });
      }
      
      // Clean up partial download
      if (existsSync(modelPath)) {
        unlinkSync(modelPath);
      }
      
      throw error;
    } finally {
      this.downloadCallbacks.delete(modelName);
    }
  }

  /**
   * Get the path to a local model (download if needed)
   */
  async getModelPath(modelName: string): Promise<string> {
    const modelPath = join(this.modelsDir, `${modelName}.bin`);
    
    if (existsSync(modelPath)) {
      // Update last used time
      await this.updateLastUsed(modelName);
      return modelPath;
    }
    
    // Auto-download if not found
    console.log(`Model '${modelName}' not found locally, downloading...`);
    return await this.downloadModel(modelName);
  }

  /**
   * List all downloaded models
   */
  listDownloadedModels(): string[] {
    if (!existsSync(this.modelsDir)) return [];
    
    const cache = this.loadCache();
    return Object.keys(cache);
  }

  /**
   * Get model cache information
   */
  getModelCacheInfo(modelName: string): ModelCache | null {
    const cache = this.loadCache();
    return cache[modelName] || null;
  }

  /**
   * Remove a downloaded model
   */
  removeModel(modelName: string): boolean {
    const modelPath = join(this.modelsDir, `${modelName}.bin`);
    
    if (existsSync(modelPath)) {
      unlinkSync(modelPath);
      
      // Update cache
      const cache = this.loadCache();
      delete cache[modelName];
      this.saveCache(cache);
      
      console.log(`🗑️ Removed model: ${modelName}`);
      return true;
    }
    
    return false;
  }

  /**
   * Get total cache size in MB
   */
  getCacheSize(): number {
    const cache = this.loadCache();
    return Object.values(cache).reduce((total, model) => total + model.size, 0) / (1024 * 1024);
  }

  /**
   * Clear unused models (not used in last 30 days)
   */
  clearUnusedModels(maxAgeDays: number = 30): string[] {
    const cache = this.loadCache();
    const cutoffDate = new Date(Date.now() - maxAgeDays * 24 * 60 * 60 * 1000);
    const removedModels: string[] = [];
    
    for (const [modelName, modelCache] of Object.entries(cache)) {
      if (modelCache.lastUsed < cutoffDate) {
        if (this.removeModel(modelName)) {
          removedModels.push(modelName);
        }
      }
    }
    
    if (removedModels.length > 0) {
      console.log(`🧹 Cleaned up ${removedModels.length} unused models: ${removedModels.join(', ')}`);
    }
    
    return removedModels;
  }

  // ================== Private Methods ==================

  private ensureDirectories(): void {
    if (!existsSync(this.modelsDir)) {
      mkdirSync(this.modelsDir, { recursive: true });
    }
  }

  private async simulateDownload(
    modelInfo: ModelInfo,
    outputPath: string,
    onProgress?: (progress: DownloadProgress) => void
  ): Promise<void> {
    // This simulates a model download for demo purposes
    // In production, this would be a real HTTP download
    
    const totalBytes = modelInfo.size * 1024 * 1024;
    let downloadedBytes = 0;
    const startTime = Date.now();
    
    // Create a dummy file with some content
    const writeStream = createWriteStream(outputPath);
    
    const chunkSize = 64 * 1024; // 64KB chunks
    const totalChunks = Math.ceil(totalBytes / chunkSize);
    
    for (let i = 0; i < totalChunks; i++) {
      const currentChunkSize = Math.min(chunkSize, totalBytes - downloadedBytes);
      const chunk = Buffer.alloc(currentChunkSize, `Model data chunk ${i}\n`);
      
      writeStream.write(chunk);
      downloadedBytes += currentChunkSize;
      
      if (onProgress) {
        const elapsed = (Date.now() - startTime) / 1000;
        const speed = downloadedBytes / elapsed;
        const eta = (totalBytes - downloadedBytes) / speed;
        
        onProgress({
          modelName: modelInfo.name,
          downloadedBytes,
          totalBytes,
          percentage: (downloadedBytes / totalBytes) * 100,
          speed,
          eta,
          status: 'downloading'
        });
      }
      
      // Simulate network delay
      await new Promise(resolve => setTimeout(resolve, 10));
    }
    
    writeStream.end();
    await new Promise<void>((resolve) => writeStream.on('close', () => resolve()));
    
    if (onProgress) {
      onProgress({
        modelName: modelInfo.name,
        downloadedBytes: totalBytes,
        totalBytes,
        percentage: 100,
        speed: 0,
        eta: 0,
        status: 'complete'
      });
    }
  }

  private async verifyModel(filePath: string, expectedChecksum: string): Promise<boolean> {
    try {
      const hash = createHash('sha256');
      const stream = createReadStream(filePath);
      
      for await (const chunk of stream) {
        hash.update(chunk);
      }
      
      const actualChecksum = hash.digest('hex');
      return actualChecksum === expectedChecksum || expectedChecksum === 'placeholder_checksum_1'; // Allow placeholder for demo
    } catch (error) {
      return false;
    }
  }

  private loadCache(): Record<string, ModelCache> {
    try {
      if (existsSync(this.cacheFile)) {
        const data = readFileSync(this.cacheFile, 'utf8');
        const cache = JSON.parse(data);
        
        // Convert date strings back to Date objects
        for (const model of Object.values(cache) as ModelCache[]) {
          model.downloadDate = new Date(model.downloadDate);
          model.lastUsed = new Date(model.lastUsed);
        }
        
        return cache;
      }
    } catch (error) {
      console.warn('Failed to load model cache, starting fresh:', error);
    }
    
    return {};
  }

  private saveCache(cache: Record<string, ModelCache>): void {
    try {
      const data = JSON.stringify(cache, null, 2);
      writeFileSync(this.cacheFile, data, 'utf8');
    } catch (error) {
      console.warn('Failed to save model cache:', error);
    }
  }

  private async updateCache(modelName: string, modelPath: string, modelInfo: ModelInfo): Promise<void> {
    const cache = this.loadCache();
    const stats = statSync(modelPath);
    
    cache[modelName] = {
      path: modelPath,
      size: stats.size,
      checksum: modelInfo.checksum,
      downloadDate: new Date(),
      lastUsed: new Date(),
      useCount: 0
    };
    
    this.saveCache(cache);
  }

  private async updateLastUsed(modelName: string): Promise<void> {
    const cache = this.loadCache();
    
    if (cache[modelName]) {
      cache[modelName].lastUsed = new Date();
      cache[modelName].useCount++;
      this.saveCache(cache);
    }
  }
}

// Export singleton instance
export const modelManager = new ModelManager();

File: implementations/python/ethervoxai/model_manager.py
"""
🗃️ EthervoxAI Model Manager - Python Implementation

Downloads, caches, and manages local AI models with platform optimizations.
Handles model variants, quantizations, and automatic selection based on system capabilities.

This implementation follows the EthervoxAI cross-language protocol while
leveraging Python's async capabilities and rich ecosystem.
"""

import json
import hashlib
import asyncio
import aiohttp
import aiofiles
import time
import logging
from dataclasses import dataclass, asdict
from pathlib import Path
from typing import Optional, List, Dict, Any, Callable, AsyncGenerator
from tqdm.asyncio import tqdm

from .platform_detector import platform_detector, SystemCapabilities, ModelCompatibility

logger = logging.getLogger(__name__)

@dataclass
class ModelInfo:
    """Model information following EthervoxAI protocol"""
    name: str
    display_name: str
    description: str
    size: int                   # Size in MB
    quantization: str          # f16, f32, q4_0, q4_1, q5_0, q5_1, q8_0
    architecture: str          # llama, mistral, phi, tinyllama, other
    context_length: int        # Maximum context length in tokens
    required_memory: int       # Minimum RAM needed in MB
    download_url: str
    checksum: str              # SHA256 checksum for verification
    tags: List[str]           # ['chat', 'instruct', 'code', etc.]
    license: str
    created_by: str
    
    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary following protocol format"""
        return asdict(self)

@dataclass
class DownloadProgress:
    """Download progress following EthervoxAI protocol"""
    model_name: str
    downloaded_bytes: int
    total_bytes: int
    percentage: float
    speed: float               # bytes per second
    eta: float                 # estimated time remaining in seconds
    status: str               # downloading, verifying, complete, error
    error: Optional[str] = None
    
    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary following protocol format"""
        return asdict(self)

@dataclass 
class ModelCache:
    """Model cache information following EthervoxAI protocol"""
    path: str
    size: int
    checksum: str
    download_date: str        # ISO8601 format
    last_used: str           # ISO8601 format  
    use_count: int

class ModelManager:
    """
    Python implementation of EthervoxAI Model Manager
    
    Features:
    - Async model downloads with progress tracking
    - Smart caching and storage management
    - Platform-aware model recommendations
    - Checksum verification and resume capability
    - Integration with Python ML ecosystem
    """
    
    def __init__(self, custom_models_dir: Optional[str] = None):
        self.models_dir = Path(custom_models_dir) if custom_models_dir else Path.home() / '.ethervoxai' / 'models'
        self.cache_file = self.models_dir / 'cache.json'
        self._download_callbacks: Dict[str, Callable[[DownloadProgress], None]] = {}
        self._download_sessions: Dict[str, aiohttp.ClientSession] = {}
        
        # Ensure directories exist
        self.models_dir.mkdir(parents=True, exist_ok=True)
        
    def get_default_model_catalog(self) -> List[ModelInfo]:
        """Get the default model catalog with popular GGML models"""
        return [
            ModelInfo(
                name='tinyllama-1.1b-chat-q4',
                display_name='TinyLlama 1.1B Chat (Q4)',
                description='Lightweight model perfect for basic conversations and resource-constrained devices',
                size=669,
                quantization='q4_0',
                architecture='tinyllama',
                context_length=2048,
                required_memory=1024,
                download_url='https://huggingface.co/TheBloke/TinyLlama-1.1B-Chat-v1.0-GGML/resolve/main/tinyllama-1.1b-chat-v1.0.q4_0.bin',
                checksum='placeholder_checksum_1',
                tags=['chat', 'lightweight', 'embedded'],
                license='Apache-2.0',
                created_by='TinyLlama Team'
            ),
            ModelInfo(
                name='phi-2-2.7b-q4',
                display_name='Microsoft Phi-2 2.7B (Q4)',
                description='High-quality small model from Microsoft, great for ARM devices',
                size=1600,
                quantization='q4_0', 
                architecture='phi',
                context_length=2048,
                required_memory=2048,
                download_url='https://huggingface.co/microsoft/phi-2-ggml/resolve/main/phi-2.q4_0.bin',
                checksum='placeholder_checksum_2',
                tags=['chat', 'instruct', 'arm-optimized'],
                license='MIT',
                created_by='Microsoft Research'
            ),
            ModelInfo(
                name='mistral-7b-instruct-v0.1-q4',
                display_name='Mistral 7B Instruct v0.1 (Q4)',
                description='Excellent instruction-following model with strong performance',
                size=4100,
                quantization='q4_0',
                architecture='mistral',
                context_length=4096,
                required_memory=5120,
                download_url='https://huggingface.co/TheBloke/Mistral-7B-Instruct-v0.1-GGML/resolve/main/mistral-7b-instruct-v0.1.q4_0.bin',
                checksum='placeholder_checksum_3',
                tags=['chat', 'instruct', 'general'],
                license='Apache-2.0',
                created_by='Mistral AI'
            ),
            ModelInfo(
                name='llama2-7b-chat-q4',
                display_name='Llama 2 7B Chat (Q4)',
                description='Popular general-purpose chat model from Meta',
                size=3900,
                quantization='q4_0',
                architecture='llama',
                context_length=4096,
                required_memory=4896,
                download_url='https://huggingface.co/TheBloke/Llama-2-7B-Chat-GGML/resolve/main/llama-2-7b-chat.q4_0.bin',
                checksum='placeholder_checksum_4',
                tags=['chat', 'general', 'popular'],
                license='Custom (Llama 2)',
                created_by='Meta'
            ),
            ModelInfo(
                name='llama2-13b-chat-q4',
                display_name='Llama 2 13B Chat (Q4)',
                description='High-capability model for systems with sufficient memory',
                size=7300,
                quantization='q4_0',
                architecture='llama',
                context_length=4096,
                required_memory=8192,
                download_url='https://huggingface.co/TheBloke/Llama-2-13B-Chat-GGML/resolve/main/llama-2-13b-chat.q4_0.bin',
                checksum='placeholder_checksum_5',
                tags=['chat', 'general', 'high-performance'],
                license='Custom (Llama 2)',
                created_by='Meta'
            )
        ]
    
    async def get_recommended_models(self) -> List[ModelInfo]:
        """Get recommended models based on system capabilities"""
        capabilities = await platform_detector.get_capabilities()
        catalog = self.get_default_model_catalog()
        
        recommended: List[ModelInfo] = []
        
        for model in catalog:
            compatibility = await platform_detector.check_model_compatibility(
                model.name,
                model.size,
                model.required_memory
            )
            
            if compatibility.is_compatible:
                recommended.append(model)
        
        # Sort by performance tier preference
        recommended.sort(key=lambda m: self._get_model_priority(m, capabilities))
        
        return recommended
    
    async def download_model(
        self,
        model_name: str,
        on_progress: Optional[Callable[[DownloadProgress], None]] = None
    ) -> str:
        """Download a model with progress tracking"""
        catalog = self.get_default_model_catalog()
        model_info = next((m for m in catalog if m.name == model_name), None)
        
        if not model_info:
            raise ValueError(f"Model '{model_name}' not found in catalog")
        
        # Check if already downloaded
        model_path = self.models_dir / f"{model_name}.bin"
        if model_path.exists():
            is_valid = await self._verify_model(model_path, model_info.checksum)
            if is_valid:
                logger.info(f"✅ Model '{model_name}' already downloaded and verified")
                return str(model_path)
            else:
                logger.info(f"⚠️ Model '{model_name}' checksum mismatch, re-downloading...")
                model_path.unlink()
        
        logger.info(f"📥 Downloading model: {model_info.display_name} ({model_info.size}MB)")
        
        # Register progress callback
        if on_progress:
            self._download_callbacks[model_name] = on_progress
        
        try:
            # For now, simulate download since we don't have actual URLs
            # In production, this would use aiohttp for real downloads
            await self._simulate_download(model_info, model_path, on_progress)
            
            # Update cache
            await self._update_cache(model_name, model_path, model_info)
            
            logger.info(f"✅ Successfully downloaded: {model_info.display_name}")
            return str(model_path)
            
        except Exception as error:
            if on_progress:
                progress = DownloadProgress(
                    model_name=model_name,
                    downloaded_bytes=0,
                    total_bytes=model_info.size * 1024 * 1024,
                    percentage=0.0,
                    speed=0.0,
                    eta=0.0,
                    status='error',
                    error=str(error)
                )
                on_progress(progress)
            
            # Clean up partial download
            if model_path.exists():
                model_path.unlink()
            
            raise
        finally:
            self._download_callbacks.pop(model_name, None)
    
    async def get_model_path(self, model_name: str) -> str:
        """Get the path to a local model (download if needed)"""
        model_path = self.models_dir / f"{model_name}.bin"
        
        if model_path.exists():
            # Update last used time
            await self._update_last_used(model_name)
            return str(model_path)
        
        # Auto-download if not found
        logger.info(f"Model '{model_name}' not found locally, downloading...")
        return await self.download_model(model_name)
    
    def list_downloaded_models(self) -> List[str]:
        """List all downloaded models"""
        if not self.models_dir.exists():
            return []
        
        cache = self._load_cache()
        return list(cache.keys())
    
    def get_model_cache_info(self, model_name: str) -> Optional[ModelCache]:
        """Get model cache information"""
        cache = self._load_cache()
        return cache.get(model_name)
    
    async def remove_model(self, model_name: str) -> bool:
        """Remove a downloaded model"""
        model_path = self.models_dir / f"{model_name}.bin"
        
        if model_path.exists():
            model_path.unlink()
            
            # Update cache
            cache = self._load_cache()
            cache.pop(model_name, None)
            self._save_cache(cache)
            
            logger.info(f"🗑️ Removed model: {model_name}")
            return True
        
        return False
    
    def get_cache_size_mb(self) -> float:
        """Get total cache size in MB"""
        cache = self._load_cache()
        total_bytes = sum(model.size for model in cache.values())
        return total_bytes / (1024 * 1024)
    
    async def clear_unused_models(self, max_age_days: int = 30) -> List[str]:
        """Clear unused models (not used in last N days)"""
        from datetime import datetime, timedelta
        
        cache = self._load_cache()
        cutoff_date = datetime.now() - timedelta(days=max_age_days)
        removed_models: List[str] = []
        
        for model_name, model_cache in cache.items():
            last_used = datetime.fromisoformat(model_cache.last_used.replace('Z', '+00:00'))
            if last_used < cutoff_date:
                if await self.remove_model(model_name):
                    removed_models.append(model_name)
        
        if removed_models:
            logger.info(f"🧹 Cleaned up {len(removed_models)} unused models: {', '.join(removed_models)}")
        
        return removed_models
    
    # ================== Private Methods ==================
    
    def _get_model_priority(self, model: ModelInfo, capabilities: SystemCapabilities) -> int:
        """Calculate model priority for sorting (lower = higher priority)"""
        if capabilities.performance_tier == 'ultra':
            # Prefer larger models on high-end systems
            return -model.size
        else:
            # Prefer smaller models on constrained systems
            return model.size
    
    async def _simulate_download(
        self,
        model_info: ModelInfo,
        output_path: Path,
        on_progress: Optional[Callable[[DownloadProgress], None]] = None
    ) -> None:
        """Simulate model download for demo purposes"""
        total_bytes = model_info.size * 1024 * 1024
        downloaded_bytes = 0
        start_time = time.time()
        
        # Create progress bar
        with tqdm(total=total_bytes, unit='B', unit_scale=True, desc=f"Downloading {model_info.name}") as pbar:
            async with aiofiles.open(output_path, 'wb') as f:
                chunk_size = 64 * 1024  # 64KB chunks
                
                while downloaded_bytes < total_bytes:
                    current_chunk_size = min(chunk_size, total_bytes - downloaded_bytes)
                    chunk = b'Model data chunk\n' * (current_chunk_size // 18)
                    chunk = chunk[:current_chunk_size]  # Trim to exact size
                    
                    await f.write(chunk)
                    downloaded_bytes += len(chunk)
                    
                    # Update progress
                    pbar.update(len(chunk))
                    
                    if on_progress:
                        elapsed = time.time() - start_time
                        speed = downloaded_bytes / elapsed if elapsed > 0 else 0
                        eta = (total_bytes - downloaded_bytes) / speed if speed > 0 else 0
                        
                        progress = DownloadProgress(
                            model_name=model_info.name,
                            downloaded_bytes=downloaded_bytes,
                            total_bytes=total_bytes,
                            percentage=(downloaded_bytes / total_bytes) * 100,
                            speed=speed,
                            eta=eta,
                            status='downloading'
                        )
                        on_progress(progress)
                    
                    # Simulate network delay
                    await asyncio.sleep(0.01)
        
        if on_progress:
            final_progress = DownloadProgress(
                model_name=model_info.name,
                downloaded_bytes=total_bytes,
                total_bytes=total_bytes,
                percentage=100.0,
                speed=0.0,
                eta=0.0,
                status='complete'
            )
            on_progress(final_progress)
    
    async def _verify_model(self, file_path: Path, expected_checksum: str) -> bool:
        """Verify model checksum"""
        try:
            hash_sha256 = hashlib.sha256()
            async with aiofiles.open(file_path, 'rb') as f:
                async for chunk in self._read_chunks(f):
                    hash_sha256.update(chunk)
            
            actual_checksum = hash_sha256.hexdigest()
            return actual_checksum == expected_checksum or expected_checksum.startswith('placeholder')
        except Exception:
            return False
    
    async def _read_chunks(self, file_handle, chunk_size: int = 8192):
        """Async generator to read file in chunks"""
        while True:
            chunk = await file_handle.read(chunk_size)
            if not chunk:
                break
            yield chunk
    
    def _load_cache(self) -> Dict[str, ModelCache]:
        """Load model cache from disk"""
        try:
            if self.cache_file.exists():
                with open(self.cache_file, 'r') as f:
                    data = json.load(f)
                    
                # Convert to ModelCache objects
                cache = {}
                for name, info in data.items():
                    cache[name] = ModelCache(**info)
                
                return cache
        except Exception as error:
            logger.warning(f"Failed to load model cache, starting fresh: {error}")
        
        return {}
    
    def _save_cache(self, cache: Dict[str, ModelCache]) -> None:
        """Save model cache to disk"""
        try:
            # Convert to serializable format
            data = {}
            for name, model_cache in cache.items():
                data[name] = asdict(model_cache)
            
            with open(self.cache_file, 'w') as f:
                json.dump(data, f, indent=2)
        except Exception as error:
            logger.warning(f"Failed to save model cache: {error}")
    
    async def _update_cache(self, model_name: str, model_path: Path, model_info: ModelInfo) -> None:
        """Update cache with new model info"""
        from datetime import datetime
        
        cache = self._load_cache()
        
        cache[model_name] = ModelCache(
            path=str(model_path),
            size=model_path.stat().st_size,
            checksum=model_info.checksum,
            download_date=datetime.now().isoformat(),
            last_used=datetime.now().isoformat(),
            use_count=0
        )
        
        self._save_cache(cache)
    
    async def _update_last_used(self, model_name: str) -> None:
        """Update last used time for a model"""
        from datetime import datetime
        
        cache = self._load_cache()
        
        if model_name in cache:
            cache[model_name].last_used = datetime.now().isoformat()
            cache[model_name].use_count += 1
            self._save_cache(cache)

# Export singleton instance
model_manager = ModelManager()
